File: frontend/src/types/hotspot.ts
export type HotspotType =
  | 'industrial_thermal_source'
  | 'mining_thermal_source'
  | 'natural_fire'
  | 'unknown';

export type ActivityStatus = 'new' | 'recurring' | 'persistent' | 'under_review';

export type Severity = 'low' | 'medium' | 'high' | 'critical';

export type HotspotStatus = 'active' | 'resolved' | 'monitoring';

export interface OSMContextFeature {
  id: string;
  featureType: string;
  name: string | null;
  distanceKm: number;
}

export interface Hotspot {
  id: string;
  latitude: number;
  longitude: number;
  type: HotspotType;  // Raw FIRMS telemetry type
  brightness: number;
  confidence: number; // NASA FIRMS confidence
  severity: Severity;
  timestamp: string;
  facilityId: string | null;
  status: HotspotStatus;
  // Phase 6 ML Prediction fields
  mlType?: HotspotType;
  mlConfidence?: number;
  modelVersion?: string;
  mlExplanation?: string | Record<string, number>;
  // ESA WorldCover 10m land-cover context
  landCoverClass?: number;
  landCoverName?: string;
  frp?: number | null;
  // Source Persistence & Operational Context
  sourceObsCount?: number;
  firstSeen?: string;
  lastSeen?: string;
  maxFrp?: number | null;
  activityStatus?: ActivityStatus;
  osmContext?: OSMContextFeature[];
}

export const HOTSPOT_COLORS: Record<HotspotType, string> = {
  industrial_thermal_source: '#FF4444',
  mining_thermal_source: '#FF8C00',
  natural_fire: '#10B981',
  unknown: '#A855F7',
};

export const HOTSPOT_LABELS: Record<HotspotType, string> = {
  industrial_thermal_source: 'Industrial Thermal Source',
  mining_thermal_source: 'Mining Thermal Source',
  natural_fire: 'Natural Fire',
  unknown: 'Under Review',
};

export const HOTSPOT_SUB_LABELS: Record<HotspotType, string> = {
  industrial_thermal_source: 'ML: Confirmed Industrial',
  mining_thermal_source: 'ML: Confirmed Mining',
  natural_fire: 'ML: Natural Vegetation Fire',
  unknown: 'ML: Unknown / Unclassified',
};

export const ACTIVITY_STATUS_LABELS: Record<ActivityStatus, string> = {
  new: 'New Detection',
  recurring: 'Active / Recurring',
  persistent: 'Persistent Source',
  under_review: 'Under Review',
};

export const ACTIVITY_STATUS_COLORS: Record<ActivityStatus, string> = {
  new: '#38BDF8',       // Light Sky Blue
  recurring: '#F59E0B', // Amber
  persistent: '#EF4444',// Red
  under_review: '#64748B' // Slate Grey
};

File: ml-service/app/db/models/hotspot.py
"""
Hotspot database model.
Matches the frontend Hotspot TypeScript interface with additional
geographic fields for India-wide coverage.
"""
from sqlalchemy import Column, DateTime, Float, ForeignKey, Index, Integer, String
from geoalchemy2 import Geometry

from app.db.base import Base


class Hotspot(Base):
    """Thermal anomaly detection record."""

    __tablename__ = "hotspots"

    id = Column(String, primary_key=True)
    latitude = Column(Float, nullable=False)
    longitude = Column(Float, nullable=False)
    type = Column(String, nullable=False)  # raw satellite type
    brightness = Column(Float, nullable=False)  # Kelvin
    confidence = Column(Float, nullable=False)  # 0-100
    severity = Column(String, nullable=False)  # low | medium | high | critical
    timestamp = Column(DateTime(timezone=True), nullable=False)
    facility_id = Column(String, ForeignKey("facilities.id"), nullable=True)
    frp = Column(Float, nullable=True)  # Fire Radiative Power (MW)
    status = Column(String, nullable=False, default="active")  # active | resolved | monitoring

    # Geographic fields for India-wide coverage
    city = Column(String, nullable=True)
    district = Column(String, nullable=True)
    state = Column(String, nullable=True)
    country = Column(String, nullable=True, default="India")

    # Source traceability: NULL = demo/seed data, "VIIRS_SNPP_NRT" / "MODIS_NRT" = FIRMS
    source = Column(String, nullable=True)

    # ML Classification predictions
    ml_type = Column(String, nullable=True)  # industrial_thermal_source | mining_thermal_source | natural_fire | unknown
    ml_confidence = Column(Float, nullable=True)  # 0.0 to 1.0 ML confidence score
    model_version = Column(String, nullable=True, default="thermalwatch-v1")
    ml_explanation = Column(String, nullable=True)  # JSON string of feature importances / SHAP contributions

    # ESA WorldCover 10m land-cover context (supplementary, NOT an ML feature)
    land_cover_class = Column(Integer, nullable=True)  # ESA WorldCover class ID (10-100)
    land_cover_name = Column(String, nullable=True)  # Human-readable class name (e.g. "Built-up", "Cropland")

    # PostGIS geometry column: POINT(longitude latitude), SRID 4326 (WGS84)
    geometry = Column(Geometry("POINT", srid=4326), nullable=True)

    __table_args__ = (
        Index("ix_hotspots_type", "type"),
        Index("ix_hotspots_ml_type", "ml_type"),
        Index("ix_hotspots_severity", "severity"),
        Index("ix_hotspots_state", "state"),
        Index("ix_hotspots_timestamp", "timestamp"),
        Index("ix_hotspots_facility_id", "facility_id"),
        Index("ix_hotspots_geometry", "geometry", postgresql_using="gist"),
    )
